### PR TITLE
Teuchos: accept ostream precision for YAML values

### PR DESCRIPTION
**Pull request by ibaned**
As seen in gahansen/Albany@5e8ded1, the previous
code was losing too much precision and caused
Albany regression tests to fail.
Increasing to always use scientific notation
with 15 decimal places.

@trilinos/teuchos 
this is part of the solution to gahansen/Albany#95

### CODE REVIEW

File: packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp [hunk 1]
```diff
@@ -659,17 +662,11 @@ void generalWriteString(const std::string& str, std::ostream& yaml)
...
-    yaml << d;
-  }
-  else
-  {
-    yaml << std::scientific << d;
-  }
-  yaml << std::fixed;
+  std::ios saved_state(0);
+  saved_state.copyfmt(yaml);
+  yaml << std::noshowpoint << std::scientific << std::setprecision(15);
```
Comment (bartlettroscoe) [inline packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp]: It think if you want full precision, you need to write out **at least** 17 digits to ensure that you read back in the same floating point number.   See:

* https://en.wikipedia.org/wiki/Double-precision_floating-point_format

That is critical if you want a binary restart.

Comment (ibaned) [inline packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp]: Actually, I would go further and say that the only way I would feel comfortable guaranteeing binary restart is with hexadecimal encodings, e.g. C++11's `std::hexfloat`. Unfortunately, that format is not so good for human readability. This PR is a balance between accuracy and readability. The goal is to keep errors low enough that regression tests don't start failing due to XML->YAML conversion, but I really don't expect binary exactness.

**Comment by bartlettroscoe** [comment]
If you write out enough digits in decimal form, you will get binary reproducability.  But that is up to you.  Actually, the user should be able to control this on the ostream object.   You should not set this in the internal code.

**Comment by ibaned** [comment]
> If you write out enough digits in decimal form, you will get binary reproducability

I'd like to see a proof of that, but that seems like an unnecessary digression.

> the user should be able to control this on the ostream object

Thats a fair point, we can do it that way as long as people are using the `std::ostream` interface, and that is sufficient for the Albany converter, which uses that interface. However, there are higher level interfaces that accept only the file name, so we'd need to decide on some default precision. Maybe just remove all modifiers from this function, accept the C++ library's defaults for the high-level YAML interface, and move this precision request over the Albany code calling the `ostream` API ? If that sounds good, I'll try it and submit a new PR if it works.

**Comment by lxmota** [comment]
As one of the users of this by way of Albany, I would say that 17 digits would do. I am not sure that having the option to control the format of the ostream object would be essential from an end-user point of view. In this particular case less is more. We are using this for converting input files in Albany from XML to YAML, which hopefully should happen only once. Then the converter should hopefully be used infrequently.

**Comment by mhoemmen** [comment]
> I'd like to see a proof of that, but that seems like an unnecessary digression.

Guy Steele has a constructive proof somewhere.  It's constructive because he gives an algorithm.

**Comment by bartlettroscoe** [comment]
> > If you write out enough digits in decimal form, you will get binary reproducability
>
> I'd like to see a proof of that, but that seems like an unnecessary digression.

According to:

* https://ece.uwaterloo.ca/~dwharder/NumericalAnalysis/02Numerics/Double/paper.pdf

in the section "Binary to Decimal Conversion", 17 digits should allow for binary predictability.

BTW, That is a great paper.  Read it my first year as a grad student and in came in handy many times (especially to weed out numerical problems with my algorithms or implementations).  It should be required reading for anyone doing numerical computing.

**Comment by lxmota** [comment]
Ah yes, that paper. I have an old copy of it somewhere.

The relevant proof for the current discussion is for Theorem 15 on page 236: binary to decimal conversion. I had totally forgotten about this!

**Comment by ibaned** [comment]
@mhoemmen @bartlettroscoe Thank you for the papers ! At least one of them confirms the 17 digit figure. I've done what @bartlettroscoe  suggested and moved the precision setting code out of Teuchos and into the Albany code that calls Teuchos (which will construct an `ostream` and set precision to 17 decimal places). This PR is still needed just to remove the previous hardcoded precision settings.

**Comment by ibaned** [comment]
Tested and pushed as e799157

**Pull request closed without merging**